_posts/2016-02-07-coupling-in-distributed-systems.md

```diff
@@ -0,0 +1,88 @@
+---
+layout: post
+name: Coupling-in-distributed-systems
+title: 'Coupling in distributed systems'
+date: 2016-02-07 00:20:00 +00:00
+author: Felipe Fernández
+image:
+    src: /assets/img/custom/blog/coupling.png
+tags:
+- distributed systems
+- microservices
+- coupling
+
+---
+
+Coupling and cohesion are key quality indicators. We strive for systems highly cohesive and loosely coupled, but high doesn't mean pure. The same goes with functional programming, we aim for isolating and reducing side effects, but we need them unless we want a useless system. It's good to modularise our systems, so whenever those modules need to talk each other they'll effectively couple themselves. Our work is create cohesive modules so that coupling is minimised.
+
+Let's provide an example. Our system has the following structure:
+
+1. Different deployables, aka, microservices architecture.
+2. Intracommunication through [Kafka](http://kafka.apache.org/) (pub-sub messaging). No HTTP involved. 
+3. 1 Producer to N Consumers scenarios.
+4. Json for data serialization.
+
+The messages that are published and consumed in this system have a schema and it's our choice making it implicit or explicit and validating that schema at compile or runtime execution. Before of analising the trade-offs of every approach let say some words about compile vs runtime approaches.
+
+## Proofing software correctness as soon as possible
+
+I've been a user od static typed languages most of my career so I'm really biased with this topic. I've strongly believe in Lean concepts as the importance of minimising waste. At the same time I love the therapeutic ideas behind Agile, TDD or BDD about exposing the truth as soon as possible. Static types, and in the end the compiler, help me to achieve those goals.
+
+I would prefer spending my time creating tests under the [motivations](https://twitter.com/sarahmei/status/685907333889810432) of providing living documentation, easing future refactors or helping me to drive the design, more than helping catching bugs that the type system should take care of. Writing a test that checks the behaviour of a method when receives null it's a waste of time if we can make impossible to write a line of code that passes a null.
+
+Compile world is not perfect though, as it's definitively slower on development and constrains developers (someone could say that less freedom might be a nice to have in this context)
+
+## Runtime approaches
+
+Now that I've been honest with you about my compile bias I can explain different approaches and trade-offs for the schema validation problem.
+
+### Implicit schemas
+
+First runtime approach is the loosest one: using implicit schemas and trusting in the good will of producers. As nobody is checking the validity of messages before of being published into Kafka that means that consumers could blow up.
```

Comment:
`validity of messages before of being published` -> `validity of messages before being published` or `validity of messages before they are published`